pkg/vault/certificate_manager.go

```diff
@@ -0,0 +1,151 @@
+package vault
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"text/template"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/cenkalti/backoff"
+	"github.com/hashicorp/vault/api"
+)
+
+type DefaultCertificateManager struct {
+	client      *api.Client
+	Certificate *Certificate
+	lease       time.Duration
+	renew       time.Duration
+	provider    *VaultSecretsProvider
+	template    *template.Template
+	outPath     string
+}
+
+func (m DefaultCertificateManager) Run(ctx context.Context, c chan int) {
+	go func() {
+		log.Printf("renewing certificate every %s", m.renew)
+
+		renewTicks := time.Tick(m.renew)
+
+		for {
+			select {
+			case <-ctx.Done():
+				log.Infof("stopping renewal")
+				return
+			case <-renewTicks:
+				err := m.Renew(ctx)
+				if err != nil {
+					log.Errorf("error renewing secret: %s", err)
+				}
+				if err == ErrPermissionDenied || err == ErrLeaseNotFound {
+					log.Error("secret could no longer be renewed")
+					c <- 1
+					return
+				}
+				err = m.save()
+				if err != nil {
+					log.Errorf("error overwriting lease: %s", err)
+				}
+			}
+		}
+	}()
+
+}
+
```

Comment:
it feels like there's a fair amount of generic stuff in this file such as the Run, renew auth token, revoke self etc, I think it would be nice if we could share this code between the certificate and credentials managers